nuxeo-features/nuxeo-search/nuxeo-search-rest-api/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/search/QueryExecutor.java

```diff
@@ -0,0 +1,378 @@
+import org.nuxeo.ecm.automation.jaxrs.io.documents.PaginableDocumentModelListImpl;
+import org.nuxeo.ecm.automation.server.jaxrs.RestOperationException;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentModelList;
+import org.nuxeo.ecm.core.api.SortInfo;
+import org.nuxeo.ecm.core.api.impl.SimpleDocumentModel;
+import org.nuxeo.ecm.core.api.model.PropertyNotFoundException;
+import org.nuxeo.ecm.platform.query.api.PageProvider;
+import org.nuxeo.ecm.platform.query.api.PageProviderDefinition;
+import org.nuxeo.ecm.platform.query.api.PageProviderService;
+import org.nuxeo.ecm.platform.query.nxql.CoreQueryDocumentPageProvider;
+import org.nuxeo.ecm.restapi.server.jaxrs.adapters.SearchAdapter;
+import org.nuxeo.ecm.webengine.model.impl.AbstractResource;
+import org.nuxeo.ecm.webengine.model.impl.ResourceTypeImpl;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * @since 8.3
+ */
+public abstract class QueryExecutor extends AbstractResource<ResourceTypeImpl> {
+
+    public static final String NXQL = "NXQL";
+
+    public static final String QUERY = "query";
+
+    public static final String PAGE_SIZE = "pageSize";
+
+    public static final String CURRENT_PAGE_INDEX = "currentPageIndex";
+
+    public static final String MAX_RESULTS = "maxResults";
+
+    public static final String SORT_BY = "sortBy";
+
+    public static final String SORT_ORDER = "sortOrder";
+
+    public static final String ORDERED_PARAMS = "queryParams";
+
+    public static final String CURRENT_USERID_PATTERN = "$currentUser";
+
+    public static final String CURRENT_REPO_PATTERN = "$currentRepository";
+
+    public enum QueryParams {
+        PAGE_SIZE, CURRENT_PAGE_INDEX, MAX_RESULTS, SORT_BY, SORT_ORDER, ORDERED_PARAMS, QUERY
+    }
+
+    public enum LangParams {
+        NXQL
+    }
+    protected EnumMap<QueryParams, String> queryParametersMap;
+
+    protected EnumMap<LangParams, String> langPathMap;
+
+    protected PageProviderService pageProviderService;
+
+    private static final Log log = LogFactory.getLog(SearchObject.class);
+
+    public void initExecutor() {
+        pageProviderService = Framework.getLocalService(PageProviderService.class);
+        // Query Enum Parameters Map
+        queryParametersMap = new EnumMap<>(QueryParams.class);
+        queryParametersMap.put(QueryParams.PAGE_SIZE, PAGE_SIZE);
+        queryParametersMap.put(QueryParams.CURRENT_PAGE_INDEX, CURRENT_PAGE_INDEX);
+        queryParametersMap.put(QueryParams.MAX_RESULTS, MAX_RESULTS);
+        queryParametersMap.put(QueryParams.SORT_BY, SORT_BY);
+        queryParametersMap.put(QueryParams.SORT_ORDER, SORT_ORDER);
+        queryParametersMap.put(QueryParams.QUERY, QUERY);
+        queryParametersMap.put(QueryParams.ORDERED_PARAMS, ORDERED_PARAMS);
+        // Lang Path Enum Map
+        langPathMap = new EnumMap<>(LangParams.class);
+        langPathMap.put(LangParams.NXQL, NXQL);
```

Comment:
The usual pattern we follow is to give a constructor to the enum to store its associated string as a field, and use a getter to get it. No need for an EnumMap.